compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala

```diff
@@ -1373,406 +1412,324 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
+         *
+         * this property is useful, later, when we need to make statements
+         * conditional based on the chosen overload.
+         */
+        var overloadNum = 0
+        def subTree(ctor: SplitSecondaryJSCtor): ConstructorTree = {
+          val low = overloadNum + 1
+          val subtrees = ctorToChildren(ctor.sym).map(subTree(_))
+          overloadNum += 1
+          new ConstructorTree(low, overloadNum, ctor, subtrees)
         }
+
+        ctorToChildren(primaryCtor.sym).map(subTree(_))
       }
-    }
 
-    private def buildJSConstructorDef(dispatch: js.JSMethodDef,
-        ctors: List[js.MethodDef])(
-        implicit pos: Position): js.JSMethodDef = {
+      /* prepare overload dispatch for all constructors.
+       * as a side-product, we retrieve the capture parameters.
+       */
+      val (exports, jsClassCaptures) = {
+        val exports = List.newBuilder[Exported]
+        val jsClassCaptures = List.newBuilder[js.ParamDef]
+
+        def add(sym: Symbol, params: List[Symbol], overloadNum: Int) = {
+          val (e, c) = genJSClassCtorDispatch(sym, params, overloadNum)
+          exports += e
+          jsClassCaptures ++= c
+        }
 
-      val js.JSMethodDef(_, dispatchName, dispatchArgs, dispatchRestParam, dispatchResolution) =
-        dispatch
+        def addTree(tree: ConstructorTree): Unit = {
+          add(tree.ctor.sym, tree.ctor.params, tree.overloadNum)
+          tree.subCtors.foreach(addTree(_))
+        }
 
-      val jsConstructorBuilder = mkJSConstructorBuilder(ctors)
+        add(primaryCtor.sym, primaryCtor.params, 0)
+        ctorTrees.foreach(addTree(_))
 
-      val overloadIdent = freshLocalIdent("overload")
+        (exports.result(), jsClassCaptures.result())
+      }
 
-      // Section containing the overload resolution and casts of parameters
-      val overloadSelection = mkOverloadSelection(jsConstructorBuilder,
-        overloadIdent, dispatchResolution)
+      val (formalArgs, restParam, overloadDispatchBody) =
+        genOverloadDispatch(JSName.Literal("constructor"), exports, jstpe.IntType)
 
-      /* Section containing all the code executed before the call to `this`
-       * for every secondary constructor.
-       */
-      val prePrimaryCtorBody =
-        jsConstructorBuilder.mkPrePrimaryCtorBody(overloadIdent)
+      val overloadVar = js.VarDef(freshLocalIdent("overload"), NoOriginalName,
+          jstpe.IntType, mutable = false, overloadDispatchBody)
 
-      val primaryCtorBody = jsConstructorBuilder.primaryCtorBody
+      val callTree = genJSClassCtorStats(overloadVar.ref, primaryCtor, ctorTrees)
 
-      /* Section containing all the code executed after the call to this for
-       * every secondary constructor.
-       */
-      val postPrimaryCtorBody =
-        jsConstructorBuilder.mkPostPrimaryCtorBody(overloadIdent)
+      // TODO: Not having js.Undefined()  passes IR checking but makes Emitter fail.
```

Comment:
@sjrd to you know from the top of your head what is going on here? I didn't have time to minimize / investigate yet.

Reply:
A JavaScript constructor function can return a value. If that value is a primitive, then the result of `new` is the originally allocated object. However, if it is an object, that value becomes the result of the `new` operator. It's a terrible spec, but it is what it is.

Therefore, our constructors in the IR also return an `any` (like all `JSMethodDef`s), and the last expression is the return value of the constructor. To make sure that `new`s of our `class`es always return the actual instance, we make them always end in `undefined`, which is a primitive value.

So the `js.Undefined()` is required for correct semantics of our language. It is also correct that the IR checker lets something else through, because that is valid IR code with dedicated semantics. It's not OK for the Emitter to crash if given that, though. Do you remember what was the issue, exactly?

Reply:
The exception is:

```
[error] java.lang.IllegalArgumentException: Illegal tree in JSDesugar.pushLhsInto():
[error] lhs = ReturnFromFunction
[error] rhs = JSSuperConstructorCall(List()) of class class org.scalajs.ir.Trees$JSSuperConstructorCall
```

So I think the Emitter is right here: We still need a return value. But `JSSuperConstructorCall` is a statement.

Reply:
Filed as #4491.